### PR TITLE
Dashboard with New Structure

### PR DESCRIPTION
**Pull request by Plypy**
Though this branch names ID-93, it covers more than EJS to JADE.

In this large PR, there are mainly:
1. All templates is converted to JADE, and other related helper code in controller level is changed. This would make the MVC separation clearer, logic simpler. Like, `headAppend` will be handled in templates, instead of in controller level sources.
2. Project structure is changed, I refered Keystone's organization style. Concept of `system-modules` is abandoned, since it won't make the project more clearer, but a little cumbersome.
3. Variables are passed via function parameters through `modules.exports`, instead of globals. 
4. Lots of packages are updated, like Formage(2.x -> 3.x), Express(3.x -> 4.x), Nodemailer(0.3.x -> 1.x), etc.
5. Changed some modules'. Now, email-verification only have one ID(UUIDv4, instead of predicatable psuedo random string) and the routes related to emails is updated as well, `/profile-email` -> `/profile/email`.

And more minor things I can't remember for now.

Dashboard in current state can correctly function basic needs, however, some bugs are still remained, like using a email address to login would cause a error. I'll tackle these things later with tests.

**Caveat**: after applying this PR, you need to update your conf.js and currently user modules should be disabled.

### CODE REVIEW

**Comment by robbyoconnor** [comment]
@Plypy, one thing that makes me uncomfortable merging this -- is that it completely breaks things -- how do you plan on handling the user modules? I want a migration path in place 

@elliottwilliams -- do you have similar concerns?

**Comment by robbyoconnor** [comment]
I am not opposed to keeping this open and having you continue work on this and add commits to this PR and adding back user-modules or having **SOME** way of keeping them in place...Don't just remove them willy-nilly...we need a migration path in place...this is the same stuff that makes the AngularJS community uneasy and the same is happening here...I want GSoC to stabilize, not de-stabilize the codebase!

There is no deadline. Sooner, rather than later.

**Comment by Plypy** [comment]
Well, if you want to remove the global dependecies, This breaking change must be made. In order to get the user-modules back. You just need to do some small modifications in user-modules, like this:

## old way

```
var app = common.app;
app.get('path', function(req, res, next) {
  //something
});
```

## new way

```

exports = module.exports = function (app) {

app.get('path', function(req, res, next) {
  //something
});


};
```

This small change could bring lots of benifits, making modules separated and hence tests of modules easy.

**Comment by robbyoconnor** [comment]
Why is this not implemented then? Finish implementing this and then ping me on this PR -- I'm patient :) 

We need to support the active plugins. Make the necessary updates please so that support is possible.
- [openmrs-contrib-id-globalnavbar](https://github.com/openmrs/openmrs-contrib-id-globalnavbar) 
- [openmrs-contrib-id-sso](https://github.com/openmrs/openmrs-contrib-id-sso)
- [openmrs-contrib-id-oauth](https://github.com/openmrs/openmrs-contrib-id-oauth) 

As I mentioned...I feel **REALLY** uncomfortable merging code that will make the codebase unstable.

**Comment by robbyoconnor** [comment]
I see now -- okay pending ensuring all the views look ok -- I'll merge this

**Comment by robbyoconnor** [comment]
Would prefer these not be on my screen -- but not a blocker either
``` 
$ npm install
npm WARN package.json openmrsid@2.0.1 license should be a valid SPDX license expression
npm WARN package.json Dependency 'supertest' exists in both dependencies and devDependencies, using 'supertest@~0.8.3' from dependencies
npm WARN package.json Dependency 'should' exists in both dependencies and devDependencies, using 'should@~2.1.1' from dependencies
npm WARN package.json Dependency 'mocha' exists in both dependencies and devDependencies, using 'mocha@~1.16.2' from dependencies
npm WARN package.json Dependency 'chai' exists in both dependencies and devDependencies, using 'chai@~1.9.1' from dependencies
npm WARN unmet dependency /home/rob/openmrs/openmrs-contrib-id/node_modules/ldapjs requires bunyan@'0.22.1' but will load
npm WARN unmet dependency /home/rob/openmrs/openmrs-contrib-id/node_modules/bunyan,
npm WARN unmet dependency which is version 1.4.0
npm WARN unmet dependency /home/rob/openmrs/openmrs-contrib-id/node_modules/mocha requires jade@'0.26.3' but will load
npm WARN unmet dependency /home/rob/openmrs/openmrs-contrib-id/node_modules/jade,
npm WARN unmet dependency which is version 1.11.0
npm WARN unmet dependency /home/rob/openmrs/openmrs-contrib-id/node_modules/request/node_modules/form-data requires async@'^1.2.1' but will load
npm WARN unmet dependency /home/rob/openmrs/openmrs-contrib-id/node_modules/async,
npm WARN unmet dependency which is version 0.2.10
```

**Comment by Plypy** [comment]
I'll clear package.json later, of course.

**Comment by Plypy** [comment]
You're right, let's leave this PR open till I update the modules. There are still some minor changes need to be done in core.

**Comment by Plypy** [comment]
Small but vital fix here..

I misused base64 encoding, which is not URL-safe, in the URLs before. Luckily, or unluckily more specifically, it didn't collapse in my testing, I just found it today.

So I replaced the base64 functions with a url safe one that conforms to RFC 4648, and added a test case for them.

**Comment by robbyoconnor** [comment]
:+1:

**Comment by robbyoconnor** [comment]
I merged your PR on my local branch -- I just pushed it so I had a backup :)

File: app/routes/reset-password/index.js [hunk 1]
```diff
@@ -100,14 +103,14 @@ app.post('/reset', mid.forceLogout, function(req, res, next) {
 app.get('/reset/:id', mid.forceLogout, validate.receive,
   function(req, res, next) {
 
-  var resetId = req.params.id;
-  verification.check(resetId, function(err, valid, locals) {
+  var id = utils.urlDecode64(id);
```
Comment (robbyoconnor) [inline app/routes/reset-password/index.js:106]: Double check this...something is going on here or in urlDecode64!

Comment (robbyoconnor) [inline app/routes/reset-password/index.js:106]: I fixed it 

``` js
var id = utils.urlDecode64(req.params.id); 
```

**Comment by robbyoconnor** [comment]
Great work :)

**Pull request merged**